File: youtube/lat7-diabetes/try1.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns 
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor

diabet = pd.read_csv('ML/youtube/lat7-diabetes/diabetes.csv')
print(diabet)
print(diabet.apply(lambda x: x.dtypes))
print(diabet.apply(lambda x: x.nunique()))
print(diabet.apply(lambda x: x.duplicated().sum()))
print(diabet.info())
print(diabet.describe())
print(diabet.isnull().sum())


for col in diabet:
    print(col)
    print(diabet[col].unique())
    print('\n\n')

fig, ax = plt.subplots(2, 5, figsize=(15, 10))
sns.scatterplot(x=diabet.index, y=diabet['Pregnancies'], ax=ax[0, 0], color='blue')
sns.scatterplot(x=diabet.index, y=diabet['Glucose'], ax=ax[0, 1], color='green')
sns.scatterplot(x=diabet.index, y=diabet['BloodPressure'], ax=ax[0, 2], color='orange')
sns.scatterplot(x=diabet.index, y=diabet['SkinThickness'], ax=ax[0, 3], color='red')
sns.scatterplot(x=diabet.index, y=diabet['Insulin'], ax=ax[1, 0], color='blue')
sns.scatterplot(x=diabet.index, y=diabet['BMI'], ax=ax[1, 1], color='green')
sns.scatterplot(x=diabet.index, y=diabet['DiabetesPedigreeFunction'], ax=ax[1, 2], color='orange')
sns.scatterplot(x=diabet.index, y=diabet['Age'], ax=ax[1, 3], color='red')
sns.scatterplot(x=diabet.index, y=diabet['Outcome'], ax=ax[1, 4], color='red')
plt.tight_layout()
plt.show()

q1 = diabet['Pregnancies'].quantile(0.25)
q3 = diabet['Pregnancies'].quantile(0.75)
iqr = q3 - q1
diabet = diabet[(diabet['Pregnancies'] >= q1 - 1.5*iqr) & (diabet['Pregnancies'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['Glucose'], 25)
q3 = np.percentile(diabet['Glucose'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['Glucose'] >= q1 - 1.5*iqr) & (diabet['Glucose'] <= q3 + 1.5*iqr )]

q1 = diabet['BloodPressure'].quantile(0.25)
q3 = diabet['BloodPressure'].quantile(0.75)
iqr = q3 - q1
diabet = diabet[(diabet['BloodPressure'] >= q1 - 1.5*iqr) & (diabet['BloodPressure'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['SkinThickness'], 25)
q3 = np.percentile(diabet['SkinThickness'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['SkinThickness'] >= q1 - 1.5*iqr) & (diabet['SkinThickness'] <= q3 + 1.5*iqr )]

q1 = diabet['Insulin'].quantile(0.25)
q3 = diabet['Insulin'].quantile(0.75)
iqr = q3 - q1
diabet = diabet[(diabet['Insulin'] >= q1 - 1.5*iqr) & (diabet['Insulin'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['BMI'], 25)
q3 = np.percentile(diabet['BMI'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['BMI'] >= q1 - 1.5*iqr) & (diabet['BMI'] <= q3 + 1.5*iqr )]

q1 = diabet['DiabetesPedigreeFunction'].quantile(0.25)
q3 = diabet['DiabetesPedigreeFunction'].quantile(0.75)
iqr = q3 - q1
diabet = diabet[(diabet['DiabetesPedigreeFunction'] >= q1 - 1.5*iqr) & (diabet['DiabetesPedigreeFunction'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['Age'], 25)
q3 = np.percentile(diabet['Age'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['Age'] >= q1 - 1.5*iqr) & (diabet['Age'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['Outcome'], 25)
q3 = np.percentile(diabet['Outcome'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['Outcome'] >= q1 - 1.5*iqr) & (diabet['Outcome'] <= q3 + 1.5*iqr )]


fig, ax = plt.subplots(2, 5, figsize=(15, 10))
sns.scatterplot(x=diabet.index, y=diabet['Pregnancies'], ax=ax[0, 0], color='blue')
sns.scatterplot(x=diabet.index, y=diabet['Glucose'], ax=ax[0, 1], color='green')
sns.scatterplot(x=diabet.index, y=diabet['BloodPressure'], ax=ax[0, 2], color='orange')
sns.scatterplot(x=diabet.index, y=diabet['SkinThickness'], ax=ax[0, 3], color='red')
sns.scatterplot(x=diabet.index, y=diabet['Insulin'], ax=ax[1, 0], color='blue')
sns.scatterplot(x=diabet.index, y=diabet['BMI'], ax=ax[1, 1], color='green')
sns.scatterplot(x=diabet.index, y=diabet['DiabetesPedigreeFunction'], ax=ax[1, 2], color='orange')
sns.scatterplot(x=diabet.index, y=diabet['Age'], ax=ax[1, 3], color='red')
sns.scatterplot(x=diabet.index, y=diabet['Outcome'], ax=ax[1, 4], color='red')
plt.tight_layout()
plt.show()

numvars = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']
scaler = MinMaxScaler()
diabet[numvars] = scaler.fit_transform(diabet[numvars])

print(diabet)

x = diabet
y = diabet.pop('Outcome')

X_train, X_test, y_train, y_test = train_test_split(x,y, test_size=0.3, random_state=42, stratify=y)
print(y_train.value_counts(normalize = True))
print(y_test.value_counts(normalize = True))


k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelLR  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(modelLR, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))



modelRF  = RandomForestClassifier()
scoring = 'accuracy'
score = cross_val_score(modelRF, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))



modelMPL  = MLPClassifier()
scoring = 'accuracy'
score = cross_val_score(modelMPL, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))


modelNB  = GaussianNB()
scoring = 'accuracy'
score = cross_val_score(modelNB, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))

modelLR.fit(X_train, y_train)
y_pred_train = modelLR.predict(X_train)

train = pd.DataFrame({
    'y_train': y_train,
    'y_pred_train': y_pred_train
})
print(train)
print(train[train['y_train'] != train['y_pred_train']])

akurasi = accuracy_score(y_train, y_pred_train)
print(akurasi)
cm = confusion_matrix(y_train, y_pred_train)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred_train')
plt.ylabel('y_train')
plt.title('confusion matrix')
plt.show()




y_pred_test = modelLR.predict(X_test)

train = pd.DataFrame({
    'y_test': y_test,
    'y_pred_test': y_pred_test
})
print(train)
print(train[train['y_test'] != train['y_pred_test']])

akurasi = accuracy_score(y_test, y_pred_test)
print(akurasi)
cm = confusion_matrix(y_test, y_pred_test)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred_test')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

File: youtube/lat7-diabetes/try1.r
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(RMySQL)
library(caTools)
library(dplyr)
library(tidyr)
library(plyr)
library(glmnet)
library(lubridate)
library(tidyverse)
library(randomForest)
library(caret)
library(patchwork)
library(nnet)
library(klaR) # untuk Naive Bayes
library(rpart) # untuk Decision Trees


df <- read.csv("ML/youtube/lat7-diabetes/diabetes.csv")
print(head(df))
print(dim(df))
print(str(df))
print(summary(df))
print(colSums(is.na(df)))
print(sapply(df, class))
print(sapply(df, function(x) length(unique(na.omit(x)))))

## TIDAK BISA KARENA ADANYA TABRAKAN DENGAN PACKAGE LAINNTA
# library(dplyr)
# library(tidyverse)
# library(ggplot2)
# null1 <- df %>% summarise(across(everything(), ~n_distinct(.[!is.na(.)])))
# print(as.data.frame(null1))

for (col in colnames(df)){
    print(col)
    print(unique(df[[col]]))
    cat('\n\n')
}

# p1 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Pregnancies), color = "blue") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p2 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Glucose), color = "green") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p3 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = BloodPressure), color = "orange") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal() 
# p4 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = SkinThickness), color = "red") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p5 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Insulin), color = "orange") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal() 
# p6 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = BMI), color = "red") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p7 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = DiabetesPedigreeFunction), color = "orange") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal() 
# p8 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Age), color = "red") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p9 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Outcome), color = "red") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()

# # # Menggabungkan plot menggunakan patchwork
# plot <- p1+p2+p3+p4+p5+p6+p7+p8+p9
# # plot <- (p1 | p2) / (p3 | p4)
# print(plot)


# q1 <- quantile(df$Pregnancies, 0.25)
# q3 <- quantile(df$Pregnancies, 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, Pregnancies >= q1 - 1.5 * iqr & Pregnancies <= q3 + 1.5 * iqr)

# q1 <- quantile(df$Glucose, 0.25)
# q3 <- quantile(df$Glucose, 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, Glucose >= q1 - 1.5 * iqr & Glucose <= q3 + 1.5 * iqr)

# q1 <- quantile(df$BloodPressure , 0.25)
# q3 <- quantile(df$BloodPressure , 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, BloodPressure  >= q1 - 1.5 * iqr & BloodPressure  <= q3 + 1.5 * iqr)

# q1 <- quantile(df$SkinThickness , 0.25)
# q3 <- quantile(df$SkinThickness , 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, SkinThickness  >= q1 - 1.5 * iqr & SkinThickness  <= q3 + 1.5 * iqr)

# q1 <- quantile(df$Insulin  , 0.25)
# q3 <- quantile(df$Insulin  , 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, Insulin   >= q1 - 1.5 * iqr & Insulin   <= q3 + 1.5 * iqr)

# q1 <- quantile(df$BMI , 0.25)
# q3 <- quantile(df$BMI , 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, BMI  >= q1 - 1.5 * iqr & BMI  <= q3 + 1.5 * iqr)

# q1 <- quantile(df$DiabetesPedigreeFunction , 0.25)
# q3 <- quantile(df$DiabetesPedigreeFunction , 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, DiabetesPedigreeFunction  >= q1 - 1.5 * iqr & DiabetesPedigreeFunction  <= q3 + 1.5 * iqr)

# q1 <- quantile(df$Age, 0.25)
# q3 <- quantile(df$Age, 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, Age >= q1 - 1.5 * iqr & Age <= q3 + 1.5 * iqr)

# q1 <- quantile(df$Outcome , 0.25)
# q3 <- quantile(df$Outcome , 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, Outcome  >= q1 - 1.5 * iqr & Outcome  <= q3 + 1.5 * iqr)

# p1 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Pregnancies), color = "blue") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p2 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Glucose), color = "green") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p3 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = BloodPressure), color = "orange") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal() 
# p4 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = SkinThickness), color = "red") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p5 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Insulin), color = "orange") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal() 
# p6 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = BMI), color = "red") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p7 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = DiabetesPedigreeFunction), color = "orange") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal() 
# p8 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Age), color = "red") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()
# p9 <- ggplot(df) +
#     geom_point(aes(x = rownames(df), y = Outcome), color = "red") +
#     facet_grid(. ~ ., scales = "free") +
#     theme_minimal()

# # # Menggabungkan plot menggunakan patchwork
# plot <- p1+p2+p3+p4+p5+p6+p7+p8+p9
# # plot <- (p1 | p2) / (p3 | p4)
# print(plot)


# print(head(df))

# numvars <- c('Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age')
# scaler <- preProcess(df[numvars], method = "range")
# df[numvars] <- predict(scaler, df[numvars])
# print(head(df))

# x <- subset(df, select = -Outcome)
# y <- as.factor(df$Outcome)

# set.seed(42)
# index <- createDataPartition(y, p = 0.7, list = TRUE)
# print(head(index))


# x_train <- x[index$Resample1, ]
# x_test <- x[-index$Resample1, ]
# y_train <- y[index$Resample1]
# y_test <- y[-index$Resample1]


# # split <- sample.split(y, SplitRatio = 0.7)
# # # x_train <- x[split, ]
# # # x_test <- x[!split, ]
# # # y_train <- y[split]
# # # y_test <- y[!split]

# # # Membuat data pelatihan
# # x_train <- subset(x, split == TRUE)
# # y_train <- subset(y, split == TRUE)
# # # Membuat data pengujian
# # x_test <- subset(x, split == FALSE)
# # y_test <- subset(y, split == FALSE)



# ctrl <- trainControl(method="cv", number=5)
# modelrf <- train(x_train, y_train, method="rf", trControl=ctrl)
# print('accuracy mean')
# print(modelrf$result)
# print(mean(modelrf$result$Accuracy))
# cat('\n')

# model <- train(x_train, y_train, method="nnet", trControl=ctrl)
# print('accuracy mean')
# print(model$result)
# print(mean(model$result$Accuracy))
# cat('\n')

# model <- train(x_train, y_train, method="rpart", trControl=ctrl)
# print('accuracy mean')
# print(model$result)
# print(mean(model$result$Accuracy))
# cat('\n')

# model <- train(x_train, y_train, method="nb", trControl=ctrl)
# print('accuracy mean')
# print(model$result)
# print(mean(model$result$Accuracy))
# cat('\n')

# y_pred_train <- predict(modelrf, x_train)
# df_pred <- data.frame(
#     y_train = y_train,
#     y_pred_train = y_pred_train
# )
# print(head(df_pred))

# df_pred <- filter(df_pred, y_train != y_pred_train)
# print(head(df_pred))

# # Calculate accuracy
# accuracy <- confusionMatrix(data = y_pred_train, reference = y_train)$overall['Accuracy']
# print(accuracy)

# cat('\n\n\n\n')
# # Calculate confusion matrix
# print('Confusion Matrix')
# cm <- confusionMatrix(data = y_pred_train, reference = y_train)
# print(cm)




# y_pred_test <- predict(modelrf, x_test)
# df_pred <- data.frame(
#     y_test = y_test,
#     y_pred_test = y_pred_test
# )
# print(head(df_pred))

# df_pred <- filter(df_pred, y_test != y_pred_test)
# print(head(df_pred))

# # Calculate accuracy
# accuracy <- confusionMatrix(data = y_pred_test, reference = y_test)$overall['Accuracy']
# print(accuracy)

# cat('\n\n\n\n')
# # Calculate confusion matrix
# print('Confusion Matrix')
# cm <- confusionMatrix(data = y_pred_test, reference = y_test)
# print(cm)
